packages/ember-auto-import/ts/analyzer-syntax.ts

```diff
@@ -0,0 +1,249 @@
+import { ReadStream } from 'fs';
+
+export interface LiteralImportSyntax {
+  isDynamic: boolean;
+  specifier: string;
+}
+
+export interface TemplateImportSyntax {
+  isDynamic: boolean;
+  // these are the string parts of the template literal. The first one always
+  // comes before the first expression.
+  cookedQuasis: string[];
+  // for each of the expressions in between the cookedQuasis, this is an
+  // optional hint for what to name the expression that goes there. It's
+  // optional because in general there may not be an obvious name, but in
+  // practice there often is, and we can aid debuggability by using names that
+  // match the original code.
+  expressionNameHints: (string | null)[];
+}
+
+export type ImportSyntax = LiteralImportSyntax | TemplateImportSyntax;
+
+const marker = 'eaimeta';
+
+export function serialize(imports: ImportSyntax[]): string {
+  let tokens = [];
+  for (let imp of imports) {
+    if ('specifier' in imp) {
+      tokens.push(imp.isDynamic ? 1 : 0);
+      tokens.push(imp.specifier);
+    } else {
+      tokens.push(imp.isDynamic ? 3 : 2);
+      tokens.push(imp.cookedQuasis);
+      tokens.push(imp.expressionNameHints);
+    }
+  }
+  return `${marker}${JSON.stringify(tokens).slice(1, -1)}${marker}`;
+}
+
+export function deserialize(source: ReadStream): Promise<ImportSyntax[]> {
+  let deserializer = new Deserializer(source);
+  return deserializer.output;
+}
+
+class Deserializer {
+  private state:
+    | {
+        // we're looking for the start marker
+        type: 'finding-start';
+      }
+    | {
+        type: 'start-partial-match';
+        // how many codepoints of the marker were present at the end of the
+        // previous chunk (to handle a marker that splits across chunks)
+        partialMatch: number;
+      }
+    | {
+        // we're looking for the end marker
+        type: 'finding-end';
+        // the meta we've read so far
+        meta: string[];
+      }
+    | {
+        type: 'end-partial-match';
+        // the meta we've read so far
+        meta: string[];
+        // how many codepoints of the marker were present at the end of the
+        // previous chunk (to handle a marker that splits across chunks)
+        partialMatch: number;
+      }
+    | {
+        type: 'done';
+        meta: string;
+      } = {
+    type: 'finding-start',
+  };
+
+  output: Promise<ImportSyntax[]>;
+  private resolve: (result: ImportSyntax[]) => void;
+  private reject: (err: any) => void;
+
+  constructor(private source: ReadStream) {
+    let r: (result: ImportSyntax[]) => void, e: (err: any) => void;
+    this.output = new Promise<ImportSyntax[]>((resolve, reject) => {
+      r = resolve;
+      e = reject;
+    });
+    this.resolve = r!;
+    this.reject = e!;
```

Comment:
do we need to observe `stream.finished(stream, options, finishedCallback)` ?

Reply:
This method only accepts fs.ReadStream, not a generic readable. ReadStream always emits close.